aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -59,15 +60,9 @@ SparseCsrTensorImpl::SparseCsrTensorImpl(
 void SparseCsrTensorImpl::resize_and_clear_(
     const int64_t nnz_size,
     IntArrayRef size) {
-  // call crow_indices().options() here since the struct contructor calls the
-  // tensor constructor with args for device specific init.
-  auto empty_crow_indices = at::empty(size[0] + 1, crow_indices().options());
-  auto empty_col_indices = at::empty(nnz_size, col_indices().options());
-  auto empty_values = at::empty(nnz_size, values().options());
-
-  crow_indices_ = empty_crow_indices;
-  col_indices_ = empty_col_indices;
-  values_ = empty_values;
+  at::native::resize_output(crow_indices_, size[0] + 1);
+  at::native::resize_output(col_indices_, nnz_size);
+  at::native::resize_output(values_, nnz_size);
```

Comment:
IIUC, using `resize_output` will trigger this warning:
https://github.com/pytorch/pytorch/blob/0a72904ab4a2eef45fab33b3b7ed79a9ebfa2d60/aten/src/ATen/native/Resize.cpp#L15-L24
when indices/values have elements. Do we want this warning? Is there any reason not to use the `resize_` method as
```
crow_indices_.resize_(size[0] + 1);
...
```
?

Reply:
I was looking the entire code, and this function `resize_and_clear_` is always used internally after the creation of a new tensor.

For example, https://github.com/pytorch/pytorch/blob/fe18144618624db92b849b467d2bfa51d99746f7/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L242

 I think this is correct, if you are trying to resize existing sparse tensor you have to be very careful and IMO this is an unsafe operation  because it's not only the sizes that should maintain CSR's  properties invariance but also the values, see some examples here:  
https://github.com/pytorch/pytorch/blob/2660dc28e53556b93a5fa860e8cd175e30a1a19d/test/test_sparse_csr.py#L51

Moreover,    `resize_output` is used because the checks and  also because  avoids redispatch for cpu and cuda. Let me know your thoughts @pearu.

Reply:
OK, using `resize_output` for new tensors only sounds reasonable.

Reply:
I also am curious about this. So in a future PyTorch release the plan is to require that the sizes already be correct? What would stop us from requiring this today?

Reply:
To be consistent we should add a set of asserts for the CSR invariants as it is in the factory functions, let me add this.

Reply:
IIUC, here the `resize_output` is called on tensors that have just been created and have no sizes defined.
If this the case (@aocsa , can you confirm?) then @mruberry , your question is based on a different assumption (that sizes are defined) and hence is irrelevant here.
If this is not the case, then what are the actual sizes and why would these not be correct already?

Reply:
Following @pearu's point, it may be more appropriate here to use internal asserts to verify the sizes are expected, and then call resize_ directly

Reply:
Yes, `resize_output` is called on tensors that have just been created. The sizes will be correct if they maintain the CSR invariants related to sizes/dim.   To be clear, these are the following CSR invariants, which can be grouped in two about sizes/dim and about the values:

## CSR invariants (summarized here gh-57211)

### CSR Invariants about sizes/dim: 
> Issue https://github.com/pytorch/pytorch/issues/57211
- `len(size) == 2` 

> Checks between line 64-80: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/SparseCsrTensor.cpp
- `(crow_indices.dim == col_indices.dim == values.dim) == 1` 
- `len(crow_indices) == size[0] + 1`
- `len(col_indices) == len(values) == nnz`


### CSR Invariants about values
> Check https://github.com/pytorch/pytorch/blob/46a32e075c0bdb54e3e2a08a370d785ee5c5a6cd/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L60 
- `crow_indices[0] == 0`
 
> Issue https://github.com/pytorch/pytorch/issues/56856
- `col_indices[crow_indices[i]:crow_indices[i+1]]`
- `values[crow_indices[i]:crow_indices[i+1]]`
- `crow_indices[-1] == nnz`

As you can see there are a couple of issues to cover the remainder invariants (`crow_indices[-1] == nnz` and `len(size) == 2` gh-56856 and gh-57211).  IMO, these checks should be added after solving these issues  not necessarily only in this function nor in this PR. So `resize_output` is still  a suitable fix to solve issue gh-56696.

cc @pearu, @mruberry